File: js/hqc/src/kem.js
/**
 * HQC KEM: KeyGen, Encapsulation, Decapsulation.
 *
 * Uses SHAKE-256 (via Node.js crypto) for hashing and seed expansion.
 * Implements Fujisaki-Okeyama transform for CCA security.
 */

import { createHash, randomBytes } from 'crypto';
import {
  vectAdd, vectMul, vectResize, vectToBytes, vectFromBytes,
  vectSetBit, vectEqual, vectAlloc,
} from './gf2.js';
import { tensorEncode, tensorDecode } from './tensor.js';
import {
  SEED_BYTES, SHARED_SECRET_BYTES,
  G_FCT_DOMAIN, H_FCT_DOMAIN, K_FCT_DOMAIN,
} from './params.js';

/**
 * SHAKE-256 seed expander.
 * Wraps a SHAKE-256 instance for sequential reads.
 */
class SeedExpander {
  constructor(seed) {
    this._hash = createHash('shake256', { outputLength: 1 << 20 });
    this._hash.update(seed);
    this._buf = this._hash.digest();
    this._pos = 0;
  }

  /**
   * Read n bytes from the expander.
   * @param {number} n
   * @returns {Uint8Array}
   */
  read(n) {
    if (this._pos + n > this._buf.length) {
      // Need to extend - create a new expander with more output
      // This should never happen in practice with 1MB buffer
      throw new Error('SeedExpander: buffer exhausted');
    }
    const out = new Uint8Array(this._buf.buffer, this._buf.byteOffset + this._pos, n);
    this._pos += n;
    return out;
  }
}

/**
 * SHAKE-256 hash function.
 * @param {Uint8Array[]} inputs - data to hash
 * @param {number} outputLen - desired output length
 * @returns {Uint8Array}
 */
function shake256(inputs, outputLen) {
  const h = createHash('shake256', { outputLength: outputLen });
  for (const input of inputs) {
    h.update(input);
  }
  return new Uint8Array(h.digest());
}

/**
 * Generate a random vector of n bits using the seed expander.
 * @param {SeedExpander} se
 * @param {number} n
 * @returns {Uint32Array}
 */
function vectSetRandom(se, n) {
  const nWords = Math.ceil(n / 32);
  const nBytes = nWords * 4;
  const buf = se.read(nBytes);
  const v = vectFromBytes(buf, nWords);
  const rem = n & 31;
  if (rem !== 0) {
    v[nWords - 1] &= (1 << rem) - 1;
  }
  return v;
}

/**
 * Generate a random vector of n bits with exactly 'weight' bits set.
 * Uses rejection sampling from the seed expander.
 * @param {SeedExpander} se
 * @param {number} n
 * @param {number} weight
 * @returns {Uint32Array}
 */
function vectSetRandomFixedWeight(se, n, weight) {
  const nWords = Math.ceil(n / 32);
  const v = new Uint32Array(nWords);
  const positions = new Uint32Array(weight);

  for (let i = 0; i < weight; i++) {
    // eslint-disable-next-line no-constant-condition
    while (true) {
      const buf = se.read(4);
      let pos = buf[0] | (buf[1] << 8) | (buf[2] << 16) | (buf[3] << 24);
      pos = (pos >>> 0) % n;

      // Check for duplicates (rejection sampling)
      let duplicate = false;
      for (let j = 0; j < i; j++) {
        if (positions[j] === pos) {
          duplicate = true;
          break;
        }
      }
      if (!duplicate) {
        positions[i] = pos;
        break;
      }
    }
  }

  for (let i = 0; i < weight; i++) {
    vectSetBit(v, positions[i]);
  }

  return v;
}

/**
 * Compute d = SHAKE256(H_domain || m), 64 bytes.
 * @param {Uint8Array} m
 * @returns {Uint8Array}
 */
function computeD(m) {
  return shake256([new Uint8Array([H_FCT_DOMAIN]), m], SHARED_SECRET_BYTES);
}

/**
 * Compute theta = SHAKE256(G_domain || m || pk || d), SEED_BYTES bytes.
 * @param {Uint8Array} m
 * @param {Uint8Array} pk
 * @param {Uint8Array} d
 * @returns {Uint8Array}
 */
function computeTheta(m, pk, d) {
  return shake256(
    [new Uint8Array([G_FCT_DOMAIN]), m, pk, d],
    SEED_BYTES,
  );
}

/**
 * Compute ss = SHAKE256(K_domain || m || u_bytes || v_bytes), 64 bytes.
 * @param {Uint8Array} m
 * @param {Uint32Array} u
 * @param {Uint32Array} v
 * @param {import('./params.js').HQCParams} p
 * @returns {Uint8Array}
 */
function computeSS(m, u, v, p) {
  return shake256(
    [
      new Uint8Array([K_FCT_DOMAIN]),
      m,
      vectToBytes(u, p.vecNSizeBytes),
      vectToBytes(v, p.vecN1N2SizeBytes),
    ],
    SHARED_SECRET_BYTES,
  );
}

/**
 * PKE Encrypt.
 * @param {Uint8Array} m - message
 * @param {Uint8Array} theta - randomness seed
 * @param {Uint8Array} pk - public key
 * @param {import('./params.js').HQCParams} p
 * @returns {[Uint32Array, Uint32Array]} [u, v]
 */
function pkeEncrypt(m, theta, pk, p) {
  // Parse public key
  const pkSeed = pk.subarray(0, SEED_BYTES);
  const s = vectFromBytes(pk.subarray(SEED_BYTES), p.vecNSize64);

  // Generate h from pk_seed
  const pkExpander = new SeedExpander(pkSeed);
  const h = vectSetRandom(pkExpander, p.n);

  // Generate r1, r2 with weight WR and e with weight WE from theta
  const thetaExpander = new SeedExpander(theta);
  const r1 = vectSetRandomFixedWeight(thetaExpander, p.n, p.wr);
  const r2 = vectSetRandomFixedWeight(thetaExpander, p.n, p.wr);
  const e = vectSetRandomFixedWeight(thetaExpander, p.n, p.we);

  // u = r1 + h * r2 mod (x^n - 1)
  const hr2 = vectMul(h, r2, p.n);
  let u = vectAdd(hr2, r1);
  u = vectResize(u, p.n);

  // v = encode(m) + s * r2 + e (in GF(2)^{n1*n2})
  const encoded = tensorEncode(m, p);

  // s * r2 in the ring, then truncate to n1*n2 bits
  const sr2 = vectMul(s, r2, p.n);
  const sr2Trunc = new Uint32Array(p.vecN1N2Size64);
  const copyLen = Math.min(sr2.length, p.vecN1N2Size64);
  for (let i = 0; i < copyLen; i++) sr2Trunc[i] = sr2[i];
  if (p.n1n2 % 32 !== 0 && p.vecN1N2Size64 > 0) {
    sr2Trunc[p.vecN1N2Size64 - 1] &= (1 << (p.n1n2 % 32)) - 1;
  }

  // Resize e to n1*n2
  const eResized = new Uint32Array(p.vecN1N2Size64);
  const copyE = Math.min(e.length, p.vecN1N2Size64);
  for (let i = 0; i < copyE; i++) eResized[i] = e[i];
  if (p.n1n2 % 32 !== 0 && p.vecN1N2Size64 > 0) {
    eResized[p.vecN1N2Size64 - 1] &= (1 << (p.n1n2 % 32)) - 1;
  }

  let v = vectAdd(encoded, sr2Trunc);
  v = vectAdd(v, eResized);
  v = vectResize(v, p.n1n2);

  return [u, v];
}

/**
 * Generate an HQC key pair.
 * @param {import('./params.js').HQCParams} p
 * @returns {{publicKey: Uint8Array, secretKey: Uint8Array}}
 */
export function keyGen(p) {
  // Generate random seeds
  const skSeed = randomBytes(SEED_BYTES);
  const pkSeed = randomBytes(SEED_BYTES);

  // Generate secret vectors x, y from sk_seed
  const skExpander = new SeedExpander(skSeed);
  const x = vectSetRandomFixedWeight(skExpander, p.n, p.w);
  const y = vectSetRandomFixedWeight(skExpander, p.n, p.w);

  // Generate random vector h from pk_seed
  const pkExpander = new SeedExpander(pkSeed);
  const h = vectSetRandom(pkExpander, p.n);

  // Compute s = x + h * y mod (x^n - 1)
  const hy = vectMul(h, y, p.n);
  let s = vectAdd(hy, x);
  s = vectResize(s, p.n);

  // Serialize public key: [pk_seed (40 bytes)] [s (VecNSizeBytes bytes)]
  const pk = new Uint8Array(p.pkSize);
  pk.set(skSeed.subarray(0, 0)); // no-op, just showing structure
  pk.set(new Uint8Array(pkSeed), 0);
  pk.set(vectToBytes(s, p.vecNSizeBytes), SEED_BYTES);

  // Serialize secret key: [sk_seed (40 bytes)] [pk (PKSize bytes)]
  const sk = new Uint8Array(p.skSize);
  sk.set(new Uint8Array(skSeed));
  sk.set(pk, SEED_BYTES);

  return { publicKey: pk, secretKey: sk };
}

/**
 * Encapsulate a shared secret using the public key.
 * @param {Uint8Array} pk
 * @param {import('./params.js').HQCParams} p
 * @returns {{ciphertext: Uint8Array, sharedSecret: Uint8Array}}
 */
export function encaps(pk, p) {
  // Generate random message m
  const m = randomBytes(p.vecKSizeBytes);

  // Compute d = H(m)
  const d = computeD(new Uint8Array(m));

  // Compute theta
  const theta = computeTheta(new Uint8Array(m), pk, d);

  // PKE Encrypt
  const [u, v] = pkeEncrypt(new Uint8Array(m), theta, pk, p);

  // Compute shared secret
  const ss = computeSS(new Uint8Array(m), u, v, p);

  // Serialize ciphertext: [u (VecNSizeBytes)] [v (VecN1N2SizeBytes)] [d (64 bytes)]
  const ct = new Uint8Array(p.ctSize);
  ct.set(vectToBytes(u, p.vecNSizeBytes));
  ct.set(vectToBytes(v, p.vecN1N2SizeBytes), p.vecNSizeBytes);
  ct.set(d, p.vecNSizeBytes + p.vecN1N2SizeBytes);

  return { ciphertext: ct, sharedSecret: ss };
}

/**
 * Decapsulate a shared secret from a ciphertext using the secret key.
 * @param {Uint8Array} sk
 * @param {Uint8Array} ct
 * @param {import('./params.js').HQCParams} p
 * @returns {Uint8Array} sharedSecret (64 bytes)
 */
export function decaps(sk, ct, p) {
  // Parse secret key: [sk_seed (40 bytes)] [pk (PKSize bytes)]
  const skSeed = sk.subarray(0, SEED_BYTES);
  const pk = sk.subarray(SEED_BYTES);

  // Parse ciphertext: [u (VecNSizeBytes)] [v (VecN1N2SizeBytes)] [d (64 bytes)]
  const u = vectFromBytes(ct.subarray(0, p.vecNSizeBytes), p.vecNSize64);
  const v = vectFromBytes(
    ct.subarray(p.vecNSizeBytes, p.vecNSizeBytes + p.vecN1N2SizeBytes),
    p.vecN1N2Size64,
  );
  const d = ct.subarray(p.vecNSizeBytes + p.vecN1N2SizeBytes);

  // Regenerate secret vectors x, y and sigma from sk_seed
  const skExpander = new SeedExpander(skSeed);
  /* x = */ vectSetRandomFixedWeight(skExpander, p.n, p.w); // consume x
  const y = vectSetRandomFixedWeight(skExpander, p.n, p.w);
  // Generate sigma (rejection secret)
  const sigma = skExpander.read(p.vecKSizeBytes);

  // Compute v - u * y (which is v XOR u*y since we are in GF(2))
  const uy = vectMul(u, y, p.n);

  // Truncate uy to n1*n2 bits
  const uyTrunc = new Uint32Array(p.vecN1N2Size64);
  const copyLen = Math.min(uy.length, p.vecN1N2Size64);
  for (let i = 0; i < copyLen; i++) uyTrunc[i] = uy[i];
  if (p.n1n2 % 32 !== 0 && p.vecN1N2Size64 > 0) {
    uyTrunc[p.vecN1N2Size64 - 1] &= (1 << (p.n1n2 % 32)) - 1;
  }

  const vMinusUY = vectAdd(v, uyTrunc);

  // Decode using tensor product code
  let [mPrime, ok] = tensorDecode(vMinusUY, p);
  if (!ok) {
    // Decoding failed - use sigma as rejection value
    mPrime = new Uint8Array(p.vecKSizeBytes);
    mPrime.set(sigma);
  }

  // Re-encrypt to verify
  const thetaPrime = computeTheta(mPrime, pk, d);
  const [u2, v2] = pkeEncrypt(mPrime, thetaPrime, pk, p);

  // Constant-time comparison
  const u2Trunc = vectResize(u2, p.n);
  const uOrig = vectResize(u, p.n);
  const uMatch = vectEqual(u2Trunc, uOrig);

  const v2Trunc = vectResize(v2, p.n1n2);
  const vOrig = vectResize(v, p.n1n2);
  const vMatch = vectEqual(v2Trunc, vOrig);

  const match = uMatch & vMatch;

  // Constant-time selection of message or sigma
  const mc = new Uint8Array(p.vecKSizeBytes);
  const maskOK = (0 - match) & 0xFF;     // 0xFF if match, 0x00 otherwise
  const maskFail = (0 - (1 - match)) & 0xFF; // 0x00 if match, 0xFF otherwise
  for (let i = 0; i < p.vecKSizeBytes; i++) {
    mc[i] = (mPrime[i] & maskOK) | (sigma[i] & maskFail);
  }

  // Compute shared secret
  return computeSS(mc, u, v, p);
}

File: rust/hqc/src/kem.rs
/// HQC KEM: Key Generation, Encapsulation, Decapsulation.
///
/// Implements the Fujisaki-Okeyama transform for CCA security on top of
/// the HQC PKE scheme.

use rand_core::CryptoRngCore;
use sha3::digest::{ExtendableOutput, Update, XofReader};
use sha3::Shake256;

use crate::gf2::*;
use crate::params::*;
use crate::tensor::*;

/// SHAKE256-based seed expander.
struct SeedExpander {
    reader: sha3::Shake256Reader,
}

impl SeedExpander {
    fn new(seed: &[u8]) -> Self {
        let mut hasher = Shake256::default();
        hasher.update(seed);
        SeedExpander {
            reader: hasher.finalize_xof(),
        }
    }

    fn read(&mut self, buf: &mut [u8]) {
        self.reader.read(buf);
    }
}

/// Generates a random vector of n bits using the seed expander.
fn vect_set_random(se: &mut SeedExpander, n: usize) -> Vec<u64> {
    let n_words = (n + 63) / 64;
    let n_bytes = n_words * 8;
    let mut buf = vec![0u8; n_bytes];
    se.read(&mut buf);
    let mut v = vect_from_bytes(&buf, n_words);
    let rem = n % 64;
    if rem != 0 {
        v[n_words - 1] &= (1u64 << rem) - 1;
    }
    v
}

/// Generates a random vector of n bits with exactly `weight` bits set,
/// using rejection sampling from the seed expander.
fn vect_set_random_fixed_weight(se: &mut SeedExpander, n: usize, weight: usize) -> Vec<u64> {
    let n_words = (n + 63) / 64;
    let mut v = vec![0u64; n_words];
    let mut positions: Vec<u32> = Vec::with_capacity(weight);
    let mut buf = [0u8; 4];

    for _ in 0..weight {
        loop {
            se.read(&mut buf);
            let pos = u32::from_le_bytes(buf) % (n as u32);

            // Rejection sampling for duplicates
            if !positions.contains(&pos) {
                positions.push(pos);
                break;
            }
        }
    }

    for &pos in &positions {
        vect_set_bit(&mut v, pos as usize);
    }

    v
}

/// Computes d = SHAKE256(H_domain || m), producing 64 bytes.
fn compute_d(m: &[u8]) -> Vec<u8> {
    let mut h = Shake256::default();
    h.update(&[HF_CT_DOMAIN]);
    h.update(m);
    let mut d = vec![0u8; SHARED_SECRET_BYTES];
    h.finalize_xof().read(&mut d);
    d
}

/// Computes theta = SHAKE256(G_domain || m || pk || d).
fn compute_theta(m: &[u8], pk: &[u8], d: &[u8], _p: &Params) -> Vec<u8> {
    let mut h = Shake256::default();
    h.update(&[GF_CT_DOMAIN]);
    h.update(m);
    h.update(pk);
    h.update(d);
    let mut theta = vec![0u8; SEED_BYTES];
    h.finalize_xof().read(&mut theta);
    theta
}

/// Computes ss = SHAKE256(K_domain || m || u_bytes || v_bytes).
fn compute_ss(m: &[u8], u: &[u64], v: &[u64], p: &Params) -> Vec<u8> {
    let mut h = Shake256::default();
    h.update(&[KF_CT_DOMAIN]);
    h.update(m);
    h.update(&vect_to_bytes(u, p.vec_n_size_bytes));
    h.update(&vect_to_bytes(v, p.vec_n1n2_size_bytes));
    let mut ss = vec![0u8; SHARED_SECRET_BYTES];
    h.finalize_xof().read(&mut ss);
    ss
}

/// PKE Encrypt: returns (u, v).
fn pke_encrypt(m: &[u8], theta: &[u8], pk: &[u8], p: &Params) -> (Vec<u64>, Vec<u64>) {
    // Parse public key
    let pk_seed = &pk[..SEED_BYTES];
    let s = vect_from_bytes(&pk[SEED_BYTES..], p.vec_n_size64);

    // Generate h from pk_seed
    let mut pk_expander = SeedExpander::new(pk_seed);
    let h = vect_set_random(&mut pk_expander, p.n);

    // Generate r1, r2, e from theta
    let mut theta_expander = SeedExpander::new(theta);
    let r1 = vect_set_random_fixed_weight(&mut theta_expander, p.n, p.wr);
    let r2 = vect_set_random_fixed_weight(&mut theta_expander, p.n, p.wr);
    let e = vect_set_random_fixed_weight(&mut theta_expander, p.n, p.we);

    // u = r1 + h * r2 mod (x^n - 1)
    let hr2 = vect_mul(&h, &r2, p.n);
    let u = vect_add(&hr2, &r1);
    let u = vect_resize(&u, p.n);

    // v = encode(m) + s * r2 + e (in GF(2)^{n1*n2})
    let encoded = tensor_encode(m, p);

    // s * r2 in the ring, then truncate to n1*n2 bits
    let sr2 = vect_mul(&s, &r2, p.n);
    let mut sr2_trunc = vec![0u64; p.vec_n1n2_size64];
    let copy_len = sr2.len().min(p.vec_n1n2_size64);
    sr2_trunc[..copy_len].copy_from_slice(&sr2[..copy_len]);
    if p.n1n2 % 64 != 0 && p.vec_n1n2_size64 > 0 {
        sr2_trunc[p.vec_n1n2_size64 - 1] &= (1u64 << (p.n1n2 % 64)) - 1;
    }

    // Resize e to n1*n2
    let mut e_resized = vec![0u64; p.vec_n1n2_size64];
    let copy_len = e.len().min(p.vec_n1n2_size64);
    e_resized[..copy_len].copy_from_slice(&e[..copy_len]);
    if p.n1n2 % 64 != 0 && p.vec_n1n2_size64 > 0 {
        e_resized[p.vec_n1n2_size64 - 1] &= (1u64 << (p.n1n2 % 64)) - 1;
    }

    let v = vect_add(&encoded, &sr2_trunc);
    let v = vect_add(&v, &e_resized);
    let v = vect_resize(&v, p.n1n2);

    (u, v)
}

/// Generates an HQC key pair.
/// Returns (public_key, secret_key).
pub fn key_gen(p: &Params, rng: &mut impl CryptoRngCore) -> (Vec<u8>, Vec<u8>) {
    // Generate random seeds
    let mut sk_seed = vec![0u8; SEED_BYTES];
    rng.fill_bytes(&mut sk_seed);
    let mut pk_seed = vec![0u8; SEED_BYTES];
    rng.fill_bytes(&mut pk_seed);

    // Generate secret vectors x, y from sk_seed
    let mut sk_expander = SeedExpander::new(&sk_seed);
    let x = vect_set_random_fixed_weight(&mut sk_expander, p.n, p.w);
    let y = vect_set_random_fixed_weight(&mut sk_expander, p.n, p.w);

    // Generate random vector h from pk_seed
    let mut pk_expander = SeedExpander::new(&pk_seed);
    let h = vect_set_random(&mut pk_expander, p.n);

    // Compute s = x + h * y mod (x^n - 1)
    let hy = vect_mul(&h, &y, p.n);
    let s = vect_add(&hy, &x);
    let s = vect_resize(&s, p.n);

    // Serialize public key: [pk_seed (40 bytes)] [s (VecNSizeBytes bytes)]
    let mut pk = vec![0u8; p.pk_size];
    pk[..SEED_BYTES].copy_from_slice(&pk_seed);
    let s_bytes = vect_to_bytes(&s, p.vec_n_size_bytes);
    pk[SEED_BYTES..SEED_BYTES + s_bytes.len()].copy_from_slice(&s_bytes);

    // Serialize secret key: [sk_seed (40 bytes)] [pk (PKSize bytes)]
    let mut sk = vec![0u8; p.sk_size];
    sk[..SEED_BYTES].copy_from_slice(&sk_seed);
    sk[SEED_BYTES..SEED_BYTES + pk.len()].copy_from_slice(&pk);

    (pk, sk)
}

/// Encapsulates a shared secret using the public key.
/// Returns (ciphertext, shared_secret).
pub fn encaps(pk: &[u8], p: &Params, rng: &mut impl CryptoRngCore) -> (Vec<u8>, Vec<u8>) {
    // Generate random message m
    let mut m = vec![0u8; p.vec_k_size_bytes];
    rng.fill_bytes(&mut m);

    // Compute d = H(m)
    let d = compute_d(&m);

    // Compute theta
    let theta = compute_theta(&m, pk, &d, p);

    // PKE Encrypt
    let (u, v) = pke_encrypt(&m, &theta, pk, p);

    // Compute shared secret
    let ss = compute_ss(&m, &u, &v, p);

    // Serialize ciphertext: [u (VecNSizeBytes)] [v (VecN1N2SizeBytes)] [d (64 bytes)]
    let mut ct = vec![0u8; p.ct_size];
    let u_bytes = vect_to_bytes(&u, p.vec_n_size_bytes);
    let v_bytes = vect_to_bytes(&v, p.vec_n1n2_size_bytes);
    ct[..p.vec_n_size_bytes].copy_from_slice(&u_bytes);
    ct[p.vec_n_size_bytes..p.vec_n_size_bytes + p.vec_n1n2_size_bytes].copy_from_slice(&v_bytes);
    ct[p.vec_n_size_bytes + p.vec_n1n2_size_bytes..].copy_from_slice(&d);

    (ct, ss)
}

/// Decapsulates a shared secret from a ciphertext using the secret key.
pub fn decaps(sk: &[u8], ct: &[u8], p: &Params) -> Vec<u8> {
    // Parse secret key: [sk_seed (40 bytes)] [pk (PKSize bytes)]
    let sk_seed = &sk[..SEED_BYTES];
    let pk = &sk[SEED_BYTES..];

    // Parse ciphertext: [u (VecNSizeBytes)] [v (VecN1N2SizeBytes)] [d (64 bytes)]
    let u = vect_from_bytes(&ct[..p.vec_n_size_bytes], p.vec_n_size64);
    let v = vect_from_bytes(
        &ct[p.vec_n_size_bytes..p.vec_n_size_bytes + p.vec_n1n2_size_bytes],
        p.vec_n1n2_size64,
    );
    let d = &ct[p.vec_n_size_bytes + p.vec_n1n2_size_bytes..];

    // Regenerate secret vectors x, y and sigma from sk_seed
    let mut sk_expander = SeedExpander::new(sk_seed);
    let _x = vect_set_random_fixed_weight(&mut sk_expander, p.n, p.w); // x not needed for decrypt
    let y = vect_set_random_fixed_weight(&mut sk_expander, p.n, p.w);
    // Generate sigma (rejection secret)
    let mut sigma = vec![0u8; p.vec_k_size_bytes];
    sk_expander.read(&mut sigma);

    // Compute v - u * y (= v XOR u*y since GF(2))
    let uy = vect_mul(&u, &y, p.n);

    // Truncate uy to n1*n2 bits
    let mut uy_trunc = vec![0u64; p.vec_n1n2_size64];
    let copy_len = uy.len().min(p.vec_n1n2_size64);
    uy_trunc[..copy_len].copy_from_slice(&uy[..copy_len]);
    if p.n1n2 % 64 != 0 && p.vec_n1n2_size64 > 0 {
        uy_trunc[p.vec_n1n2_size64 - 1] &= (1u64 << (p.n1n2 % 64)) - 1;
    }

    let v_minus_uy = vect_add(&v, &uy_trunc);

    // Decode using tensor product code
    let (m_prime, ok) = tensor_decode(&v_minus_uy, p);

    let m_prime = if ok { m_prime } else { sigma.clone() };

    // Re-encrypt to verify
    let theta_prime = compute_theta(&m_prime, pk, d, p);
    let (u2, v2) = pke_encrypt(&m_prime, &theta_prime, pk, p);

    // Constant-time comparison
    let u2_trunc = vect_resize(&u2, p.n);
    let u_orig = vect_resize(&u, p.n);
    let u_match = vect_equal(&u2_trunc, &u_orig);

    let v2_trunc = vect_resize(&v2, p.n1n2);
    let v_orig = vect_resize(&v, p.n1n2);
    let v_match = vect_equal(&v2_trunc, &v_orig);

    let matched = u_match & v_match;

    // Constant-time selection of message or sigma
    let mask_ok = 0u8.wrapping_sub(matched as u8); // 0xFF if match, 0x00 otherwise
    let mask_fail = 0u8.wrapping_sub((1 - matched) as u8); // 0x00 if match, 0xFF otherwise
    let mut mc = vec![0u8; p.vec_k_size_bytes];
    for i in 0..p.vec_k_size_bytes {
        mc[i] = (m_prime[i] & mask_ok) | (sigma[i] & mask_fail);
    }

    // Compute shared secret
    compute_ss(&mc, &u, &v, p)
}
